### PR TITLE
Remove php src the right way for lightweight containers.

### LINKED ISSUE #234: Remove php src the right way for lightweight containers
Hi guys,

in the same way we do for alpine packages in https://github.com/docker-library/php/pull/206... maybe we can think about removing `/usr/src/php` from the produced container image & make things easy to restore it?

Basically, I was thinking about:
- [ ] Add some `&& rm -rf /usr/src/php` [here](https://github.com/docker-library/php/blob/81ceba13187f9488f1ab25683575ac1b62fea772/7.0/zts/alpine/Dockerfile#L87)
- [ ] Create a new script named for example `docker-grab-php-source.sh` that can [manage the whole download/gpg check & so on](https://github.com/docker-library/php/blob/master/7.0/zts/alpine/Dockerfile#L48-L60)
- [ ] Like in #206 provide a mecanism to grab php source code each time we want to install some extension.

As `/usr/src/php` for php7 weights around 190M inside a running alpine image, I guess we can reduce the final image size by 10 to 20M (which is about the sizes of compressed php releases source code artifacts.).

To sum up, this can be around 25 to 50% gain for the final image size.

Also, this can be applied to non Alpine Linux images I guess, as it was tried with a non working approach in https://github.com/docker-library/php/pull/96.

ping @tianon @ncopa what do you thing about that?

### PR DESCRIPTION
**Pull request by shouze**
Fixes #234.

- I will call `update.sh` after the proposed patch have been reviewed.
- I've also included some kind of debian support.

According to @yosifkit https://github.com/docker-library/php/issues/234#issuecomment-222853089 we should obtain that kind of result (uncompressed size container images):
```
REPOSITORY          TAG                IMAGE ID            CREATED             VIRTUAL SIZE
# current size
php                 alpine             bdae0b0de098        3 days ago          191.5 MB

# sans all source
<none>              <none>             537592e5c872        53 seconds ago      44.3 MB

# keep the php src xz
<none>              <none>             cb1f5c08b4f9        50 minutes ago      55.79 MB
```

### CODE REVIEW

File: Dockerfile-alpine.template [hunk 1]
```diff
@@ -83,7 +74,9 @@ RUN set -xe \
 			| sort -u \
 	)" \
 	&& apk add --no-cache --virtual .php-rundeps $runDeps \
-	&& apk del .build-deps
+	&& apk del .build-deps \
+	&& docher-php-source delete source \
```
Comment (cjunge) [inline Dockerfile-alpine.template]: Should be ```docker-php-source delete source \```

File: Dockerfile-alpine.template [hunk 2]
```diff
@@ -83,7 +74,9 @@ RUN set -xe \
...
+	&& docher-php-source delete tarball
```
Comment (cjunge) [inline Dockerfile-alpine.template]: Should be ```docker-php-source delete tarball \```

File: Dockerfile-debian.template [hunk 3]
```diff
@@ -42,18 +44,8 @@ RUN set -xe \
...
-	&& for key in $GPG_KEYS; do \
-		gpg --keyserver ha.pool.sks-keyservers.net --recv-keys "$key"; \
-	done \
-	&& gpg --batch --verify "$PHP_FILENAME.asc" "$PHP_FILENAME" \
-	&& rm -r "$GNUPGHOME" "$PHP_FILENAME.asc" \
-	&& mkdir -p /usr/src/php \
-	&& tar -xf "$PHP_FILENAME" -C /usr/src/php --strip-components=1 \
-	&& rm "$PHP_FILENAME" \
+	&& docker-php-source download \
+	&& docher-php-source extract \
```
Comment (cjunge) [inline Dockerfile-debian.template]: Should be ```docker-php-source extract \```

File: Dockerfile-debian.template [hunk 4]
```diff
@@ -72,7 +64,9 @@ RUN set -xe \
 	&& make install \
 	&& { find /usr/local/bin /usr/local/sbin -type f -executable -exec strip --strip-all '{}' + || true; } \
 	&& make clean \
-	&& apt-get purge -y --auto-remove -o APT::AutoRemove::RecommendsImportant=false -o APT::AutoRemove::SuggestsImportant=false $buildDeps
+	&& apt-get purge -y --auto-remove -o APT::AutoRemove::RecommendsImportant=false -o APT::AutoRemove::SuggestsImportant=false $buildDeps \
+	&& docher-php-source delete source \
```
Comment (cjunge) [inline Dockerfile-debian.template]: Should be ```docker-php-source delete source \```

File: Dockerfile-debian.template [hunk 5]
```diff
@@ -72,7 +64,9 @@ RUN set -xe \
...
+	&& docher-php-source delete tarball
```
Comment (cjunge) [inline Dockerfile-debian.template]: Should be ```docker-php-source delete tarball \```

**Comment by shouze** [comment]
@cjunge thx for noticing me the typo... 🙀

**Comment by shouze** [comment]
ping @tianon @yosifkit any feedback? I can run update.sh if you will.

**Comment by yosifkit** [comment]
To simplify this I would rather leave the downloading in the Dockerfile and keep the tar.gz file around.  This will make the new script only need to do two things, extract and delete.  This will also guarantee that any source compiled against in your modules is the same that built the version of php in use and lessen the impact on php servers as well as time on users.

How much time does the extraction and deletion add to a typical `docker-php-ext-install`?  We should probably do some benchmarks beyond size.

This also has ramifications on the `docker-php-ext-*` scripts since they use `/usr/src/php/ext` to figure out what extensions are available in their 'usage' sections (see [enable](https://github.com/docker-library/php/blob/e37ca400ba0e1c34357bf06732bc77064e5a4941/docker-php-ext-enable#L13), [configure](https://github.com/docker-library/php/blob/e37ca400ba0e1c34357bf06732bc77064e5a4941/docker-php-ext-configure#L6), and [install](https://github.com/docker-library/php/blob/e37ca400ba0e1c34357bf06732bc77064e5a4941/docker-php-ext-install#L15)).  Minimally, we should extract at the start of the file and then delete with logic similar to the apk deletes (configure should not delete, install should delete, and I'm not sure about enable).

**Comment by shouze** [comment]
@yosifkit ok, will make the changes soon & report numbers about extract/delete times.

**Comment by shouze** [comment]
ping @yosifkit ok done.

About `docker-php-ext-*`:
- `enable`: nothing to do
- `configure` only extract from tarball if not already done.
- `install`: extract from tarball if not already done then delete.

The only last possible improvement I see is... if somebody decide in its inherited Dockerfile RUN to call manually at the beginning `docker-php-source extract` then `docker-php-ext-install` maybe should not delete everytime at the end and then `docker-php-source delete` should be explicitly at the end of the Dockerfile RUN.

I will make some benchmark about the added consumed time by extract.

**Comment by shouze** [comment]
@yosifkit benchmark done... on my local machine the extract step is very fast:
```shell
time tar -Jxf "/$PHP_FILENAME" -C /usr/src
real	0m 3.10s
user	0m 1.53s
sys	0m 2.48s
time rm -rf /usr/src/php
real	0m 0.52s
user	0m 0.04s
sys	0m 0.48s
```

I think we can go with this PR as is, I let you review prior to run update.sh & merge it.

**Comment by jeremyFreeAgent** [comment]
👍

**Comment by shouze** [comment]
ping @yosifkit @tianon

**Pull request closed without merging**